### ISSUE TITLE
(ec2.KeyPair): Resource handler returned message: "null"

### ISSUE THREAD

**Issue by ThePlenkov** [aws/aws-cdk#30311]
### Describe the bug

I have a following message when creating a new keypair:

```
Resource handler returned message: "null" (RequestToken: e736f836-0250-d35d-3b71-9762d9a7f44a, HandlerErrorCode: InternalFailure)
```

here is how I create it:
```
//create keypair
    const keyPairName = 'solace-keypair';
    const keyPair = new KeyPair(this, 'solaceKeypair', {
      keyPairName,
    });
```

What could go wrong?

### Expected Behavior

It's ok if it fails, I expect that returned message from the resource is not null

### Current Behavior

Currently it is hard to guess what happens because message is null

### Reproduction Steps

may be to apply mentioned code. Not sure, may be it depends on my account/permissions

### CDK CLI Version

2.142.1 (build ed4e152)

### Node.js Version

v20.12.0

### OS

Debian GNU/Linux 11 (bullseye)

### Language

TypeScript

**Comment by khushail** [comment]
Hey @ThePlenkov , it works fine for me and I could see the EC2.KeyPair being created in my account -

    const keypair = new ec2.KeyPair(this, "keypair", {
      keyPairName: "keypair",
    });
    
    
<img width="1039" alt="Screenshot 2024-05-22 at 2 00 35 PM" src="https://github.com/aws/aws-cdk/assets/117320115/21dd27cc-0c39-4504-bae4-bfa730b7067d">

**Comment by ThePlenkov** [comment]
That part I understand that this constructor should work because it's described in the documentation. What I want to say is that if something doesn't work ( may be no privilege) - then in certain cases ( as it happens for me in our enterprise service account) it may return response null - which is indeed not a good response. It would be great if someone could review this resource and check if that could happen

**Comment by khushail** [comment]
I agree with you @ThePlenkov that error should be displayed accordingly. AFAIK, if its related to credentials or something , it should come up like that.

In your case (with shared snippet), its not clear from where error is coming. To understand which part of execution is producing this error, I might have to repro that on my side and then look up the source code to identify from where `null` is being returned. I hope you understand my point of view here. So It would be great if you could share the complete code which produced this error.

**Comment by khushail** [comment]
You could also use `--debug` to see the verbose logging if that is helpful.

**Comment by ThePlenkov** [comment]
I did it different instead and I left only keypair in my stack, error is there.

```
// create keypair via cdk
import * as cdk from 'aws-cdk-lib';
import { type Construct } from 'constructs';
import { KeyPair } from 'aws-cdk-lib/aws-ec2';

export class TestKeyPairStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
    //create keypair
    new KeyPair(this, 'test-keypair', {
      keyPairName: 'test-keypair',
    });
  }
}
```

```
#!/usr/bin/env node
import 'source-map-support/register';
import * as cdk from 'aws-cdk-lib';
import { TestKeyPairStack } from '../lib/keypair';

const app = new cdk.App();

new TestKeyPairStack(app, 'test-keypair-stack', {
  stackName: process.env['STACK_NAME'],
  env: {
    account: process.env['CDK_DEPLOY_ACCOUNT'],
    region: process.env['CDK_DEPLOY_REGION'],
  },
  synthesizer: new cdk.DefaultStackSynthesizer({
    generateBootstrapVersionRule: false,
  }),
});
```

I cannot change the command we deploy with because it's in the Harness pipeline. Do you think there is a way to change it programmatically? Will cdk.json help may be? Thanks!

**Comment by ThePlenkov** [comment]
yes I already found that [debug ](https://docs.aws.amazon.com/cdk/v2/guide/cli.html) is possible to enable via cdk.json too

**Comment by ThePlenkov** [comment]
debug: true in cdk.json didn't help to have better logs in Harness

**Comment by ThePlenkov** [comment]
So I tried to deploy such a  template to my personal sandbox account and it worked. 
[test-keypair-stack.template.json](https://github.com/aws/aws-cdk/files/15419595/test-keypair-stack.template.json)

But this template fails when deploying to a private service account. I checked permission boundaries of the deployment account and it looks good - keypair is not restricted.

So it must be something else..

**Comment by ThePlenkov** [comment]
@khushail is it possible if you can try to run this code without permission to create key pair? Thanks!

**Comment by ThePlenkov** [comment]
@khushail  I can confirm that - after adding missing permissions to the deployment policy everything works now.

> Resource handler returned message: "null" (RequestToken: e736f836-0250-d35d-3b71-9762d9a7f44a, HandlerErrorCode: InternalFailure)

this is the message which is shown if the account doesn't have permissions.

To me it seems like a place for improvement to provide better message rather than null.

**Comment by khushail** [comment]
@ThePlenkov , it would be helpful if you could share which permissions you added to your account to make it work.

**Comment by ThePlenkov** [comment]
Just ec2*
